packages/cli/src/modules/workflow-index/workflow-index.service.ts

```diff
@@ -0,0 +1,165 @@
+			await this.dependencyRepository.updateDependenciesForWorkflow(workflowId, []);
+		});
+		this.eventService.on('workflow-unarchived', async ({ workflowId }) => {
+			await this.workflowRepository.findOneBy({ id: workflowId }).then(async (workflow) => {
+				if (workflow) {
+					await this.updateIndexFor(workflow);
+				}
+			});
+		});
+	}
+
+	private async buildIndex() {
+		// Get all the workflows.
+		const workflows = await this.workflowRepository.find();
+		// Build the index for each workflow.
+		for (const workflow of workflows) {
+			await this.updateIndexFor(workflow);
+		}
+	}
+
+	/**
+	 * Update the dependency index for a given workflow.
+	 *
+	 * NOTE: this should generally be handled via events, rather than called directly.
+	 * The exception is during workflow imports where it's simpler to call directly.
+	 *
+	 * @param workflow
+	 */
+	async updateIndexFor(workflow: IWorkflowBase) {
+		// TODO: input validation.
+		// Generate the dependency updates for the given workflow.
+		const dependencyUpdates: WorkflowDependency[] = [];
+
+		this.addNodeTypeDependencies(workflow, dependencyUpdates);
+		this.addCredentialDependencies(workflow, dependencyUpdates);
+		this.addWorkflowCallDependencies(workflow, dependencyUpdates);
+		this.addWebhookPathDependencies(workflow, dependencyUpdates);
+
+		const updated = await this.dependencyRepository.updateDependenciesForWorkflow(
+			workflow.id,
+			dependencyUpdates,
+		);
+		this.logger.debug(
+			`Workflow dependency index ${updated ? 'updated' : 'skipped'} for workflow ${workflow.id}`,
+		);
+	}
+
+	private addNodeTypeDependencies(
+		workflow: IWorkflowBase,
+		dependencyUpdates: WorkflowDependency[],
+	): void {
+		// Iterate over the nodes, extract node types, and add to dependencyUpdates.
+		workflow.nodes.forEach((node) => {
+			const dependency = new WorkflowDependency();
+			Object.assign(dependency, {
+				workflowId: workflow.id,
+				workflowVersionId: workflow.versionCounter,
+				dependencyType: 'nodeType',
+				dependencyKey: node.type,
+				dependencyInfo: node.id,
+				indexVersionId: 1, // TODO: figure out a better way to manage this.
+			});
+			dependencyUpdates.push(dependency);
+		});
+	}
+
+	private addCredentialDependencies(
+		workflow: IWorkflowBase,
+		dependencyUpdates: WorkflowDependency[],
+	): void {
+		workflow.nodes
+			.filter((node) => node.credentials) // Only include nodes with credentials.
+			.forEach((node) => {
+				// TODO: figure out if this is actually right, and what checks I need.
+				const credentialType = Object.keys(node.credentials!)[0];
+				const { id, name } = node.credentials![credentialType];
+				const dependency = new WorkflowDependency();
+				Object.assign(dependency, {
+					workflowId: workflow.id,
+					workflowVersionId: workflow.versionCounter,
+					dependencyType: 'credential',
+					dependencyKey: id,
+					dependencyInfo: JSON.stringify({ nodeId: node.id, credentialType, credentialName: name }),
```

Comment:
I don't think we need to store all this information. I think node ID is enough. The name can change and the index won't get updated. The type is available via the credential ID if needed. That way this column stays consistent with the other ones